src/Shared/TypeLoader.cs

```diff
@@ -317,31 +327,190 @@ internal LoadedType GetLoadedTypeByTypeName(string typeName)
                         }
                     }
 
-                    if (Interlocked.Read(ref _haveScannedPublicTypes) == 0)
+                    if (!taskHostFactoryExplicitlyRequested)
                     {
-                        lock (_lockObject)
+                        if (Interlocked.Read(ref _haveScannedPublicTypes) == 0)
                         {
-                            if (Interlocked.Read(ref _haveScannedPublicTypes) == 0)
+                            lock (_lockObject)
                             {
-                                ScanAssemblyForPublicTypes();
-                                Interlocked.Exchange(ref _haveScannedPublicTypes, ~0);
+                                if (Interlocked.Read(ref _haveScannedPublicTypes) == 0)
+                                {
+                                    ScanAssemblyForPublicTypes();
+                                    Interlocked.Exchange(ref _haveScannedPublicTypes, ~0);
+                                }
                             }
                         }
-                    }
 
-                    foreach (KeyValuePair<string, Type> desiredTypeInAssembly in _publicTypeNameToType)
-                    {
-                        // if type matches partially on its name
-                        if (typeName.Length == 0 || TypeLoader.IsPartialTypeNameMatch(desiredTypeInAssembly.Key, typeName))
+                        foreach (KeyValuePair<string, Type> desiredTypeInAssembly in _publicTypeNameToType)
                         {
-                            return desiredTypeInAssembly.Value;
+                            // if type matches partially on its name
+                            if (typeName.Length == 0 || TypeLoader.IsPartialTypeNameMatch(desiredTypeInAssembly.Key, typeName))
+                            {
+                                return new TypeInformation(new LoadedType(desiredTypeInAssembly.Value, _assemblyLoadInfo, _loadedAssembly));
+                            }
                         }
                     }
 
-                    return null;
+                    return FindTypeInformationUsingSystemReflectionMetadata(typeName);
                 });
 
-                return type != null ? new LoadedType(type, _assemblyLoadInfo, _loadedAssembly) : null;
+                return typeInfo;
+            }
+
+            private TypeInformation FindTypeInformationUsingSystemReflectionMetadata(string typeName)
+            {
```

Comment:
@sharwell told me that Assembly.LoadFrom(bytes) doesn't lock the assembly. He suggested we could potentially load the relevant assembly into a byte array, then use Assembly.LoadFrom to get its full information without changing almost any other code. I'm guessing that would be a little less efficient, but if it can simply load the assembly, that may be worth it.